src/main/java/cubicchunks/world/dependency/DependencyManager.java

```diff
@@ -0,0 +1,187 @@
+ *
+ *  Copyright (c) 2015 contributors
+ *
+ *  Permission is hereby granted, free of charge, to any person obtaining a copy
+ *  of this software and associated documentation files (the "Software"), to deal
+ *  in the Software without restriction, including without limitation the rights
+ *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ *  copies of the Software, and to permit persons to whom the Software is
+ *  furnished to do so, subject to the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be included in
+ *  all copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ *  THE SOFTWARE.
+ */
+
+package cubicchunks.world.dependency;
+
+import cubicchunks.CubicChunks;
+import cubicchunks.server.ServerCubeCache;
+import cubicchunks.server.ServerCubeCache.LoadType;
+import cubicchunks.util.CubeCoords;
+import cubicchunks.world.cube.Cube;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+/*
+ * TODO: Commenting
+ */
+public class DependencyManager {
+
+	// TODO: Make private.
+	public ServerCubeCache cubeProvider;
+
+	private Map<CubeCoords, RequiredCube> requiredMap;
+
+
+	public DependencyManager(ServerCubeCache cubeProvider) {
+		this.cubeProvider = cubeProvider;
+		this.requiredMap = new HashMap<>();
+	}
+
+
+	/**
+	 * Registers a given Dependent. All of the cubes it requires will be loaded up to the stage it needs them to be
+	 * at and it will be notified whenever one of its required cubes advances a stage.
+	 *
+	 * @param dependent The Dependent to be registered.
+	 */
+	public void register(Dependent dependent) {
+
+		// If the dependent does not have any requirements, there is nothing left to do here.
+		Collection<Requirement> requirements = dependent.getRequirements();
+
+		// Add all of the requirements.
+		for (Requirement requirement : requirements) {
+
+			// If the required cube is required by other dependents already, retrieve its RequiredCube instance,
+			// otherwise create a new one.
+			RequiredCube requiredCube = this.requiredMap.get(requirement.getCoords());
+			if (requiredCube == null) {
+
+				// If the required cube is loaded already, add it to the RequiredCube instance immediately.
+				Cube cube = this.cubeProvider.getCube(requirement.getCoords());
+				if (cube != null) {
+					requiredCube = new RequiredCube(cube);
+				}
+				// Otherwise create an empty instance.
+				else {
+					requiredCube = new RequiredCube();
```

Comment:
They are equivalent, its personal taste of mine to keep things like this separate as it increases readability (for me). A parameter being null just isn't the same as it not being there. If you want me to change this I'll do so.